### ISSUE TITLE
Guide/Universal: Describe how to pre-generate HTML files

### ISSUE THREAD

**Issue by matthew-a-thomas** [angular/angular#23024]
## I'm submitting a...
<pre>
[x] Documentation issue or request
</pre>

## Current behavior

The [current Angular Universal guide page](https://github.com/angular/angular/blob/505ae752b61a44ca41396838a49f83cfe7badb3b/aio/content/guide/universal.md) (located [here on the web](https://angular.io/guide/universal)) says of Angular Universal:

> It can also pre-generate pages as HTML files that you serve later.

However, no direction is given to accomplish that.

## Expected behavior

The guide page should either:

 * Describe how to accomplish pre-generated HTML files.
 * Or not mention the possibility of pre-generating HTML files.

Clearly, if pre-generating HTML files is a possibility, then the preferred solution would be to explain how to do that.

## What is the motivation / use case for changing the behavior?

Statically-hosting an Angular Universal application would provide all the [benefits of Angular Universal](https://angular.io/guide/universal#why-universal) and also give greater hosting flexibility.

**Comment by matthew-a-thomas** [comment]
Note this Stack Overflow answer mentions some of the things they needed to do in order to accomplish pre-generated HTML files:
https://stackoverflow.com/a/46143536/3063273

But not much detail was given there, either. This is especially hard to follow for beginners, like me.

**Comment by Thommas** [comment]
Would be nice to have a working setup relying on https://github.com/GoogleChrome/puppeteer

I'm building this setup right now. We will see how it goes.

I'm using Angular 6 rc0, I'm getting this error when following the universal setup:

```
ERROR in node_modules/@angular/common/src/location/location.d.ts(1,10): error TS2305: Module '"/frontend/node_modules/rxjs/Rx"' has no exported member 'SubscriptionLike'.
```

Might need to downgrade to 5.x before I can move forward...

**Comment by Toxicable** [comment]
It's not exactly Documentation, but you can check out our Universal Starter - https://github.com/angular/universal-starter 
Which has an example of prerendering pages.
This is the file that does it here https://github.com/angular/universal-starter/blob/master/prerender.ts

If you have any questions feel free to ping me on the Angular gitter https://gitter.im/angular/angular

**Comment by matthew-a-thomas** [comment]
@Toxicable Thank you for pointing to the Universal Starter. Perhaps that will help answer [my Stack Overflow question](https://stackoverflow.com/q/49521894/3063273). I'll take a closer look tomorrow.

**Comment by matthew-a-thomas** [comment]
I don't want to distract from this documentation request, but @Toxicable [you were right](https://stackoverflow.com/a/49534626/3063273) about suggesting a `document` property.

**Comment by matthew-a-thomas** [comment]
More on topic with this request, perhaps something like the below is all that's needed in the guide. It could be added somewhere at the end.

---

# Prerendering

[As mentioned above](https://angular.io/guide/universal#angular-universal-server-side-rendering), Angular Universal may be used to pre-generate HTML files that you serve later.

Now that you have your Angular Universal application being rendered server-side, the below steps are all that is needed to pre-generate the HTML files.

## The prerendering script

Copy this file alongside your `server.ts` file, and name it `prerender.ts`: https://github.com/angular/universal-starter/blob/master/prerender.ts

> If you're creating an entirely new application, [the Angular Universal Starter](https://github.com/angular/universal-starter) may be a great place to start. It gives you a solid starting point with an Angular Universal application which already supports both prerendering and server-side rendering.

## Webpack configuration

Modify your `webpack.server.config.js` file to add another `entry`:

```
module.exports = {
    entry: {
        server: './server.ts',
        'prerender': './prerender.ts'
    },
    ...
```

 > Adding `'prerender': './prerender.ts'` tells Webpack to use your prerendering script to generate `prerender.js`, which you'll run later.

## Compile

Run this command:

```
npm run build:universal
```

> This will compile your application and tell Webpack to generate a `dist/prerender.js` file.

## Execute prerendering script

Run this command:

```
node dist/prerender.js
```

> This command pre-generates all your configured routes, placing the pre-generated HTML files in their respective locations under `dist/browser`.

**Comment by jenniferfell** [comment]
Thanks, everyone for your feedback and helpful links. Is there any chance one of you would like to submit a PR? (never hurts to ask?)

**Comment by johnlabarge** [comment]
I personally tried to make pre-generate work with the tour of heroes universal and have run into quite a couple snags, and I'm having trouble resolving them.  I'm not an angular expert yet.  

My procedure was to steal the angular.json, prerender.ts and the static.paths.ts from the universal starter.  ( I modified the angular.json to have the correct project name and modified static.paths.ts to have a single path to '/').  

The first snag I run into is a syntax error that I guess has something to do with es6? 

```
/Users/johnlabarge/angular-universal-starter/tour/node_modules/angular-in-memory-web-api/interfaces.js:1
(function (exports, require, module, __filename, __dirname) { import { Injectable } from '@angular/core';
                                                                     ^

SyntaxError: Unexpected token {
    at new Script (vm.js:74:7)
    at createScript (vm.js:246:10)
    at Object.runInThisContext (vm.js:298:10)
    at Module._compile (internal/modules/cjs/loader.js:670:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
    at Module.require (internal/modules/cjs/loader.js:650:17)
    at require (internal/modules/cjs/helpers.js:20:18)
```

I resolve this snag by bundling all the files. 

```
"server": {
          "builder": "@angular-devkit/build-angular:server",
          "options": {
            "outputPath": "dist/server",
            "main": "src/main.server.ts",
            "tsConfig": "src/tsconfig.server.json",
  /*added this*/          "bundleDependencies": "all"
          },
```

This brings on the second snag.  This one is difficult to debug and I don't yet have a workaround - or enough background with Angular to speculate with insight about what is happening. 
``` 
Unhandled Promise rejection: Cannot read property 'subscribe' of undefined ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'subscribe' of undefined
    at new ApplicationRef (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:43305:41)
    at _createClass (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47228:24)
    at _createProviderInstance (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47190:30)
    at initNgModule (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47123:36)
    at new NgModuleRef_ (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47849:13)
    at createNgModuleRef (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47838:16)
    at Object.debugCreateNgModuleRef [as createNgModuleRef] (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:49663:16)
    at NgModuleFactory_../node_modules/@angular/core/bundles/core.umd.js.NgModuleFactory_.create (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:50380:29)
    at /Users/johnlabarge/angular-universal-starter/tour/node_modules/@angular/core/bundles/core.umd.js:4375:47
    at ZoneDelegate.invoke (/Users/johnlabarge/angular-universal-starter/tour/node_modules/zone.js/dist/zone-node.js:388:26) TypeError: Cannot read property 'subscribe' of undefined
    at new ApplicationRef (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:43305:41)
    at _createClass (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47228:24)
    at _createProviderInstance (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47190:30)
    at initNgModule (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47123:36)
    at new NgModuleRef_ (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47849:13)
    at createNgModuleRef (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:47838:16)
    at Object.debugCreateNgModuleRef [as createNgModuleRef] (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:49663:16)
    at NgModuleFactory_../node_modules/@angular/core/bundles/core.umd.js.NgModuleFactory_.create (/Users/johnlabarge/angular-universal-starter/tour/dist/server/main.js:50380:29)
    at /Users/johnlabarge/angular-universal-starter/tour/node_modules/@angular/core/bundles/core.umd.js:4375:47
    at ZoneDelegate.invoke (/Users/johnlabarge/angular-universal-starter/tour/node_modules/zone.js/dist/zone-node.js:388:26)
```
My highlevel suspicion is that this has something to do with attempts to prerender dynamic pages.  However this is not a graceful failure.  Also if that is the case, then it would probably make sense to work on a guide that more fully discusses what should/should not be attempted to be prerendered.

**Comment by johnlabarge** [comment]
I managed to get through this.  It looks like the universal starter was dated/incompatible with tour of heroes.

**Comment by jenniferfell** [comment]
Universal doc has been rewritten for v7.  Please let us know if you find errors or omissions in the new v7 doc. Thanks!
#25752

**Comment by johnlabarge** [comment]
I'm still not seeing the pre-generate steps which I think are important for deploying to a CDN.

**Comment by mwamufiya** [comment]
Given the addition of a universal schematic in 6.1. a lot of what is in the document currently gets generated for you via cli. [as seen here](https://medium.com/@MarkPieszak/angular-universal-server-side-rendering-deep-dive-dc442a6be7b7)

updating the documentation to leverage the schematic would make this process much easier and increase adoption I imagine..

**Comment by jocactus** [comment]
Thanks to @matthew-a-thomas, I was able to edit manually the `ng add @nguniversal/express-engine` to also handle the static pre-rendering.

For those interested, all the steps to do so:

1. `ng add @nguniversal/express-engine --clientProject [your project name]`

2. Create the files `prerender.ts` and `static.paths.ts` at the root level of your project next to `server.ts`

3. Copy the content of [https://github.com/angular/universal-starter/blob/master/prerender.ts](https://github.com/angular/universal-starter/blob/master/prerender.ts) in your `prerender.ts` file

4. Create your routes in your `static.paths.ts` following the example [https://github.com/angular/universal-starter/blob/master/static.paths.ts](https://github.com/angular/universal-starter/blob/master/static.paths.ts)

5. Add the prerender entry in your `webpack.server.config.js`:

```
module.exports = {
  mode: 'none',
  entry: {
    server: './server.ts',
    prerender: './prerender.ts'
  },
```

6. Add the prerender scripts in your `package.json` :

```
"scripts": {
    ...
    "build:prerender": "npm run build:client-and-server-bundles && npm run compile:server && npm run generate:prerender",
    "generate:prerender": "cd dist && node prerender",
    "serve:prerender": "cd dist/browser && http-server"
  }
```

7.  Edit the following line in your `prerender.ts`:
Change `const {AppServerModuleNgFactory, LAZY_MODULE_MAP} = require('./server/main');` (normally line 17)
By `const {AppServerModuleNgFactory, LAZY_MODULE_MAP} = require('./dist/server/main');`

8. Install the package **http-server** to serve the prerender build:
 `npm install http-server --save-dev`

You are now ready to go!
`npm run build:prerender && npm run serve:prerender`

```
Starting up http-server, serving ./
Available on:
  http://127.0.0.1:8080
  http://192.168.0.10:8080
Hit CTRL-C to stop the server
```

**Comment by wottpal** [comment]
Hi @JohanRin, I couldn't thank you more for your detailed explanation - It helped me a lot!

I have two questions regarding this issue:
* Is there already another issue which aims at simplifying the whole static-build process? I honestly think @JohanRin's solution is great but is also has way too much overhead. Generating static-pages should be a one-liner immediately after adding `@nguniversal/express-engine` IMHO.
* This one is a little more specific because I'm only building one single application under `/projects/` it would be nice to be able to place all the necessary files (webpack-config, prerender, server) just there and not in the root-directory. Currently when I'm trying to do this webpack throws a lot of errors that some modules can not be found. (I'm sorry I have never used webpack before, only Angular-CLI or Gulp).

**Comment by mbaljeetsingh** [comment]
@johanrin 
I'm getting the following error when i run `npm run build:prerender && npm run serve:prerender`. Any thoughts?

![image](https://user-images.githubusercontent.com/872762/63263435-34efc800-c2a6-11e9-8c20-e192f26da1b4.png)

**Comment by chrisguttandin** [comment]
Hi everyone, a friend just pointed me to this thread. I built a command line tool to prerender Angular apps called [angular-prerender](https://github.com/chrisguttandin/angular-prerender). My goal was to simplify all the steps discussed here and to automate them as much as possible. I hope it proves to be useful for some of you as well.

@kapunahelewong Please let me know if you think it would be a useful addition to the docs. I would be happy to provide a pull request which explains the basic usage of the command line tool.

**Comment by mbaljeetsingh** [comment]
@chrisguttandin I tried **angular-prerender** couple of weeks ago, it wan't detecting lazy loading routes. I will again give it a try.

I had a conversation with @MarkPieszak on **twitter**. They are adding prerendering to Angular Universal schematic soon.
![image](https://user-images.githubusercontent.com/872762/64903898-03202480-d6df-11e9-8e58-d4cfb43288ca.png)

**Comment by MarkPieszak** [comment]
Hey all!
We are in the process of simplifying and extending the Universal schematics to include prerendering by default (https://github.com/angular/universal/pull/1206), automatically generate `sitemap.xml` for those routes, (potentially) guess.js predicting prerenderable routes, and hopefully include a lot of other tools to prevent common gotchas and problems when creating SSR applications!

We're also trying to fix some Ivy integration issues, and have all building/bundling setup and running through `Builders`, instead of having Webpack/TSC separately.

When this is all finished we'll have a _much_ smoother & simplified SSR process.

Btw @chrisguttandin If you can shoot me an email I'd love to chat sometime, I know you had done some work with guess.js parser and it'd be great to hear about any issues you had in this area.

**Comment by kmturley** [comment]
@MarkPieszak 

Couple of questions:
- Any idea on when this might might become part of the default codebase?
- Does the new pre-render feature support rendering REST API responses to static json files?

I've used the pre-rendering feature on 5-6 sites, and it's amazing. For me it's a differentiator vs other frameworks.

I have always had to add it manually using universal-starter as a reference, and had to extend the TransferStateModuleFactory to pass REST API responses back to the prerender script in order to render static json files from APIs.

**Comment by KingDarBoja** [comment]
Hello Mark!

Regarding the second item ("_guess.js predicting prerenderable routes_"), I do think the **ng-toolkit** creator did something similar as part of the `ng-toolkit/universal` schematic, which can be seen on his [repo](https://github.com/maciejtreder/ng-toolkit/tree/master/schematics/universal/src/files).

Of course, the current version is not updated as pointed out at this [ng-toolkit issue](https://github.com/maciejtreder/ng-toolkit/issues/724) (which was made by myself) but if you look at the example repo I provided, you will see a fixed prerendering logic to automatically load app routes and generate static HTML files based on that.

Hopefully that should do the trick, unless this approach is too complex or not what you talking about.

Cheers!

**Comment by jaisonpjohn** [comment]
Finally this got merged https://github.com/angular/universal/pull/1357
Awaiting the release

**Comment by jaisonpjohn** [comment]
@johanrin I followed exactly these steps. But I am getting an error

Error: ENOENT: no such file or directory, mkdir '<Project Folder>/dist/browser/lazy/nested'


ng --version

Angular CLI: 8.3.20
Node: 12.4.0
OS: darwin x64
Angular:
...

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.803.20
@angular-devkit/core         8.3.20
@angular-devkit/schematics   8.3.20
@schematics/angular          8.3.20
@schematics/update           0.803.20
rxjs                         6.4.0

**Comment by kuncevic** [comment]
I have prerenderer configured with angular 8 project following https://github.com/angular/universal-starter a while a go that works fine.
Now I upgraded to 9 and got an error during prerender:
```
> dev@0.0.0 generate:prerender /tmp/dev/dev0
> cd dist && node prerender

Unhandled Promise rejection: Cannot read property 'moduleType' of undefined ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot read property 'moduleType' of undefined
    at /tmp/dev/dev0/dist/prerender.js:28905:118
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98965:30)
    at Object.onInvoke (/tmp/dev/dev0/dist/prerender.js:28385:33)
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98964:36)
    at Zone.run (/tmp/dev/dev0/dist/prerender.js:98722:47)
    at NgZone.module.exports.NgZone.run (/tmp/dev/dev0/dist/prerender.js:28282:28)
    at PlatformRef.module.exports.PlatformRef.bootstrapModuleFactory (/tmp/dev/dev0/dist/prerender.js:28904:23)
    at renderModuleFactory (/tmp/dev/dev0/dist/prerender.js:96261:39)
    at /tmp/dev/dev0/dist/prerender.js:292384:101
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98965:30) TypeError: Cannot read property 'moduleType' of undefined
    at /tmp/dev/dev0/dist/prerender.js:28905:118
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98965:30)
    at Object.onInvoke (/tmp/dev/dev0/dist/prerender.js:28385:33)
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98964:36)
    at Zone.run (/tmp/dev/dev0/dist/prerender.js:98722:47)
    at NgZone.module.exports.NgZone.run (/tmp/dev/dev0/dist/prerender.js:28282:28)
    at PlatformRef.module.exports.PlatformRef.bootstrapModuleFactory (/tmp/dev/dev0/dist/prerender.js:28904:23)
    at renderModuleFactory (/tmp/dev/dev0/dist/prerender.js:96261:39)
    at /tmp/dev/dev0/dist/prerender.js:292384:101
    at ZoneDelegate.invoke (/tmp/dev/dev0/dist/prerender.js:98965:30)
```

**Comment by kuncevic** [comment]
Updating my universal prerenderer v8 to v9 project, got that error:
 `No routes found. Specify routes to render using `prerender.options.routes` in angular.json.`

**Comment by kuncevic** [comment]
Was being caught by this error when upgrading my project with universal prerender from v8 to v9 by running `ng update @angular/cli@next @angular/core@next @nguniversal/express-engine@next @nguniversal/module-map-ngfactory-loader@next --force``

I did notice some universal configurations changes, my `server.ts` becomes `server.ts.bak`, `webpack.server.config.js` becomes `webpack.server.config.js.bak` etc,

So I did run `npm run prerender` got an error `Specify routes to render using `prerender.options.routes` in angular.json` which I fixed but on the second attempt running `npm run prerender` got that error:
`No routes found. Specify routes to render using prerender.options.routes in angular.json.`

Any ideas?

**Comment by dev-assassin** [comment]
> Thanks to @matthew-a-thomas, I was able to edit manually the `ng add @nguniversal/express-engine` to also handle the static pre-rendering.
> 
> For those interested, all the steps to do so:
> 
> 1. `ng add @nguniversal/express-engine --clientProject [your project name]`
> 2. Create the files `prerender.ts` and `static.paths.ts` at the root level of your project next to `server.ts`
> 3. Copy the content of https://github.com/angular/universal-starter/blob/master/prerender.ts in your `prerender.ts` file
> 4. Create your routes in your `static.paths.ts` following the example https://github.com/angular/universal-starter/blob/master/static.paths.ts
> 5. Add the prerender entry in your `webpack.server.config.js`:
> 
> ```
> module.exports = {
>   mode: 'none',
>   entry: {
>     server: './server.ts',
>     prerender: './prerender.ts'
>   },
> ```
> 
> 1. Add the prerender scripts in your `package.json` :
> 
> ```
> "scripts": {
>     ...
>     "build:prerender": "npm run build:client-and-server-bundles && npm run compile:server && npm run generate:prerender",
>     "generate:prerender": "cd dist && node prerender",
>     "serve:prerender": "cd dist/browser && http-server"
>   }
> ```
> 
> 1. Edit the following line in your `prerender.ts`:
>    Change `const {AppServerModuleNgFactory, LAZY_MODULE_MAP} = require('./server/main');` (normally line 17)
>    By `const {AppServerModuleNgFactory, LAZY_MODULE_MAP} = require('./dist/server/main');`
> 2. Install the package **http-server** to serve the prerender build:
>    `npm install http-server --save-dev`
> 
> You are now ready to go!
> `npm run build:prerender && npm run serve:prerender`
> 
> ```
> Starting up http-server, serving ./
> Available on:
>   http://127.0.0.1:8080
>   http://192.168.0.10:8080
> Hit CTRL-C to stop the server
> ```

when i run this command `npm run build:prerender && npm run serve:prerender`, I got this error.

```
The token '&&' is not a valid statement separator in this version.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : InvalidEndOfLine
```

**Comment by Blackyout** [comment]
> I managed to get through this. It looks like the universal starter was dated/incompatible with tour of heroes.

hello, how did you fix the 'subscribe' error?

**Comment by alan-agius4** [comment]
Prerendering information has been added at https://angular.io/guide/prerendering

Please feel free to a new issue if there is missing information.
